### PR TITLE
Feat/support tray icon

### PR DESCRIPTION
**Pull request by hunjixin**
program hide instead of exit when click close icon, if really want to exit, click tray icon and choose quit.
when program is hide,  double click tray icon to show (window work, ubuntu not work, others unknown),  can also click open menu in tray icon.

Resolves https://github.com/subspace/space-acres/issues/141

### CODE REVIEW

**Comment by hunjixin** [comment]
https://github.com/subspace/space-acres/issues/141

File: .gitignore
Comment (nazar-pc) [inline .gitignore:4]: Please don't, even `/.idea` should be removed from there

File: src/main.rs
Comment (nazar-pc) [inline src/main.rs:59]: We probably don't need a separator here

File: src/main.rs
Comment (nazar-pc) [inline src/main.rs:111]: Typo and order should be more logical (and matching order in the code below):
```suggestion
    ShowWindow,
    Quit,
```

File: src/main.rs
Comment (nazar-pc) [inline src/main.rs:210]: No dead code, please

File: src/main.rs
Comment (nazar-pc) [inline src/main.rs:510]: I don't think you need to convert icon to vector first, it is already in byte slice.

Also `.expect()`should be either proof why this never happens or replaced with proper error handling, see other places that already exist in the code. Same comment about other `.expect()` you have here, none of them are proofs, see https://github.com/subspace/subspace/blob/main/CONTRIBUTING.md#code-style

File: src/main.rs
Comment (nazar-pc) [inline src/main.rs:517]: It is not a "Demo System Tray" though, right?

File: src/main.rs
Comment (nazar-pc) [inline src/main.rs:529]: You can avoid custom names like `sender_c` like this:
```suggestion
        let tray = TrayIconBuilder::new()
            .with_icon(tray_img)
            .with_tooltip("Demo System Tray")
            .with_menu(build_menu())
            .build({
                let sender = sender.clone();
            
                move |tray_event| match tray_event {
                    TrayEvent::Menu(signal) => match signal {
                        TrayMenuSignal::Open => sender.input(AppInput::ShouWindow),
                        TrayMenuSignal::Quit => sender.input(AppInput::Quit),
                    },
                    TrayEvent::Tray(e) => match e {
                        ClickType::Double => sender.input(AppInput::ShouWindow),
                        _ => {}
                    },
                }
            })
            .expect("build tray icon");
```

File: src/main.rs
Comment (nazar-pc) [inline src/main.rs:630]: It is `AppStatusCode::Exit` by default, no need to set it explicitly

**Review by nazar-pc — COMMENTED** [review]
This is very promising, thank you!

If you only verified it on Windows and it doesn't yet work on Linux, then Linux changes should not be included here or PR can't be merged in that case.

> #141

https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/using-keywords-in-issues-and-pull-requests instead can be used instead, I edited original message.

Once comments are addressed we can merge Windows support and then probably have Ubuntu support in a separate PR (we'll need to look into dependencies as documented in `betrayer` to make sure correct dependencies are included in the package.

**Comment by hunjixin** [comment]
> This is very promising, thank you!
> 
> If you only verified it on Windows and it doesn't yet work on Linux, then Linux changes should not be included here or PR can't be merged in that case.
> 
> > #141
> 
> https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/using-keywords-in-issues-and-pull-requests instead can be used instead, I edited original message.
> 
> Once comments are addressed we can merge Windows support and then probably have Ubuntu support in a separate PR (we'll need to look into dependencies as documented in `betrayer` to make sure correct dependencies are included in the package.

the reason I added some linux code is twofold. First, to provide a basic support, and second, to prevent potential compilation failures if it's not handled.  I will work on the specific support in a separate pull request.

File: src/main.rs
Comment (hunjixin) [reply src/main.rs:210]: menu object must keepalive, if not, menu disappear after drop

**Review by nazar-pc — COMMENTED** [review]
> the reason I added some linux code is twofold. First, to provide a basic support, and second, to prevent potential compilation failures if it's not handled. I will work on the specific support in a separate pull request.

I understand, but ideally all PRs that are merged are ready for new release. We don't want to merge half-working or broken features. Do you think it'll be a lot of work to make it work on Linux in this PR?

**Comment by hunjixin** [comment]
> > the reason I added some linux code is twofold. First, to provide a basic support, and second, to prevent potential compilation failures if it's not handled. I will work on the specific support in a separate pull request.
> 
> I understand, but ideally all PRs that are merged are ready for new release. We don't want to merge half-working or broken features. Do you think it'll be a lot of work to make it work on Linux in this PR?

i have try ubuntu, its nearly okay, just no double click event,  for kde  menu button show as a checkbox, but work okay, no broken at last.  there are too much linux distribution, which versions need to be check?  we can leave this pr until test some main linux system.

**Comment by nazar-pc** [comment]
> i have try ubuntu, its nearly okay, just no double click event, for kde menu button show as a checkbox, but work okay, no broken at last. there are too much linux distribution, which versions need to be check? we can leave this pr until test some main linux system.

Ubuntu 22.04 and newer is what should work out of the box. Other distros and DEs are nice to have and ideally we would identify where tray icon is not supported and disable corresponding functionality.

File: .gitignore
Comment (hunjixin) [reply .gitignore]: done

File: src/main.rs
Comment (hunjixin) [reply src/main.rs]: removed

File: src/main.rs
Comment (hunjixin) [reply src/main.rs]: done

File: src/main.rs
Comment (hunjixin) [reply src/main.rs]: done

File: src/main.rs
Comment (hunjixin) [reply src/main.rs]: done

File: src/main.rs
Comment (hunjixin) [reply src/main.rs]: change to Space Acres

File: src/main.rs
Comment (hunjixin) [reply src/main.rs]: done, and change double to left click,   better compatible

File: src/main.rs
Comment (hunjixin) [reply src/main.rs]: done

**Comment by hunjixin** [comment]
i will start work for ubutu and kde. this pr not merge until complete linux work.

File: src/main.rs
Comment (nazar-pc) [inline src/main.rs:515]: Can be simplified:
```suggestion
            .with_menu(Menu::new([
                MenuItem::button("Open", AppInput::ShowWindow),
                MenuItem::button("Quit", AppInput::Quit),
            ]))
```

**Review by nazar-pc — COMMENTED** [review]
Thanks, the shape looks good overall, let me know once it is ready, I'll test it and merge.

BTW you don't need to comment "done" or similar under every comment, if they are trivial just resolve them yourself or I will after additional review. Also if you comment from "Files changed" tab you can submit multiple comments as a single review that decreases number of notifications I'm getting a lot.

File: src/main.rs
Comment (nazar-pc) [inline src/main.rs:109]: Still `Shou` rather than `Show`

File: src/main.rs
Comment (nazar-pc) [inline src/main.rs:627]: Why is `.maximize()` needed?

File: Cargo.toml
Comment (nazar-pc) [inline Cargo.toml:103]: This dependency is not used right now, did you plan to use it for something?

**Comment by hunjixin** [comment]
@nazar-pc i think this pr can resolve both windows and linux issue. linux has done minimal work, and betrayer is functioning well(change a little). It has been tested successfully on Ubuntu 22.04 and KDE Neon.

File: src/main.rs
Comment (hunjixin) [reply src/main.rs:627]: when windows is minimize, click open but do nothing.  just want to bring window to front,   hmm but this feature not work for linux,  maximize do nothing in this scenario.  do not thing it should be remove?

File: src/main.rs
Comment (hunjixin) [reply src/main.rs:565]: my idea is that if the goal is to minimize, it would be better to directly hide it in the system tray. If you think minimizing is better, I can change it back here.

File: src/main.rs
Comment (hunjixin) [reply src/main.rs:510]: hmm, I think I understand your point now. so, here it should be a statement rather than a reason.

File: Cargo.toml
Comment (nazar-pc) [inline Cargo.toml:39]: Please don't add git revisions from your forks, it will not be merged. Only upstream git revisions are accepted and even then they require comments explaining why it is necessary and when can be switched to upstream release.

Also you have a ton of whitespaces at the end of the line here.

File: src/main.rs
Comment (nazar-pc) [inline src/main.rs:503]: I see you've added macOS, are you sure it works there? Because if you're not sure we either need to confirm with someone or not enable this on macOS yet. I think @abhi3700 might be able to test it on macOS.

Also there is still a typo:
```suggestion
        #[cfg(any(target_os = "macos", target_os = "linux"))]
        let tray_img = Icon::from_png_bytes(TRAY_ICON).expect("Tray icon exists and is a valid PNG; qed");
        #[cfg(target_os = "windows")]
        let tray_img = Icon::from_resource(1, None).expect("Tray icon exists and is a valid PNG; qed");
```

File: src/main.rs
Comment (nazar-pc) [inline src/main.rs:522]: This has nothing to do with the build script though, so this proof is incorrect. What are the cases when this fails?

File: src/main.rs
Comment (nazar-pc) [reply src/main.rs:558]: That is not what I meant. You do not check if tray icon was enabled successfully. So if tray icon was not enabled correctly application will crash, but I think it shouldn't because tray icon is not a critical functionality. Instead tray icon should probably be stored in `Option<TrayIcon>` and we simply should minimize window instead of hiding in that case. Because if you hide it, you'll not be able to show it again without tray icon.

File: src/main.rs
Comment (nazar-pc) [inline src/main.rs:511]: Please read suggestions very carefully. What I suggested for previous code revision was to use `AppInput` instead of `TrayMenuSignal`, so you both simplify it here and don't need to match/map signals below.

File: src/main.rs
Comment (nazar-pc) [reply src/main.rs:620]: Maximize makes window occupy the whole screen, we don't want that, in fact our window is not resizable in the first place.

What you probably wanted to do is `.unminimize()`, but even then it only makes sense if we minimized it first, which we didn't we hid the window. So I don't understand why this is necessary. If you want to print it to the front, maybe `present` is what should be used instead (it will also show and un-minimize the window for you).

File: src/main.rs
Comment (nazar-pc) [inline src/main.rs:564]: I think this is more cleanly done with [`set_hide_on_close`](https://docs.gtk.org/gtk4/method.Window.set_hide_on_close.html), but this should be un-done if we fail to initialize tray icon as described in other comments.

File: Cargo.toml
Comment (nazar-pc) [reply Cargo.toml]: Well, if you need to bump the version of the crate you don't need to add it to `Cargo.toml`, you can just update that particular crate with `cargo update -p crate_name`

File: src/main.rs
Comment (hunjixin) [reply src/main.rs:503]: I  am not confirm it work on macos, but i can confirm without this macos will build fail

File: src/main.rs
Comment (hunjixin) [reply src/main.rs:511]: but the library request signal to be clone, and to make AppInput Clone change many relative types to be clone, and i see this type also involve not clone type in other lib.

```rs
impl<T: Clone + Send + 'static> TrayIconBuilder<T> {

    /// Attempts to create the tray icon. See the the *Platform notes* section of the Readme for more information.
    pub fn build<F>(self, callback: F) -> TrayResult<TrayIcon<T>>
        where F: FnMut(TrayEvent<T>) + Send + 'static
    {
        Ok(TrayIcon(NativeTrayIcon::new(self, callback)?))
    }
}
```

**Comment by nazar-pc** [comment]
Please don't abuse force push this much, it is hard to follow what (if anything) you're changing

File: Cargo.toml
Comment (hunjixin) [reply Cargo.toml]: kde and ubuntu button style not the same,  and wait for betrayer next release

File: Cargo.toml
Comment (nazar-pc) [inline Cargo.toml:39]: This should use full hash and have a TODO saying which upstream version to upgrade to to switch to upstream since this is a temporary solution. Also there are still a lot of whitespaces at the end of this line unnecessarily.

File: src/main.rs
Comment (nazar-pc) [inline src/main.rs:525]:
```suggestion
                warn!(%err, "Unable to create tray icon ");
```

**Review by nazar-pc — COMMENTED** [review]
Please squash and rebase against `main` and let me know when this is ready for review again, it seems close

File: Cargo.toml
Comment (hunjixin) [reply Cargo.toml:39]: comment on cargo.toml?

Comment (nazar-pc) [reply Cargo.toml:39]: Example of actionable TODOs that were eventually removed: https://github.com/subspace/subspace/pull/2587

**Comment by hunjixin** [comment]
@nazar-pc  have squash and rebase main,  and i have test on win11/ubuntu22.04/ubuntu23.10

File: src/main.rs
Comment (abhi3700) [inline src/main.rs:502]: Right now this is the problem:
image

The `from_png_bytes` function is not supported on `macOS`. And for `from_rgba` function, it requires different args. Don't know the purpose much.

Tried with this modification on top of latest commit so far:
```rust
        #[cfg(target_os = "linux")]
        let tray_img =
            Icon::from_png_bytes(TRAY_ICON).expect("Tray icon exists and is a valid PNG; qed");
        #[cfg(target_os = "macos")]
        let tray_img = Icon::from_rgba(
            TRAY_ICON.to_vec(),
            about_dialog.width() as u32,
            about_dialog.height() as u32,
        )
        .expect("Tray icon exists and is a valid PNG; qed");
```

File: src/main.rs
Comment (abhi3700) [inline src/main.rs:522]: This should be improved as suggested by clippy:
```sh
warning: you seem to be trying to use `match` for destructuring a single pattern. Consider using `if let`
   --> src/main.rs:523:35
    |
523 |                   move |tray_event| match tray_event {
    |  ___________________________________^
524 | |                     TrayEvent::Menu(signal) => match signal {
525 | |                         TrayMenuSignal::Open => sender.input(AppInput::ShowWindow),
526 | |                         TrayMenuSignal::Quit => sender.input(AppInput::Quit),
527 | |                     },
528 | |                     _ => {}
529 | |                 }
    | |_________________^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#single_match
    = note: `#[warn(clippy::single_match)]` on by default
help: try
    |
523 ~                 move |tray_event| if let TrayEvent::Menu(signal) = tray_event { match signal {
524 +                     TrayMenuSignal::Open => sender.input(AppInput::ShowWindow),
525 +                     TrayMenuSignal::Quit => sender.input(AppInput::Quit),
526 +                 } }
    |
```

**Review by abhi3700 — COMMENTED** [review]
Left my review comment.
cc @nazar-pc

File: src/main.rs
Comment (hunjixin) [reply src/main.rs:502]: i will try tomorrow, if okay in linux, i will both use from rgba. but why use dialog width and height

Comment (abhi3700) [reply src/main.rs:502]: > but why use dialog width and height
Idk, it's asked in `from_rgba` function, so I placed some window's dimension. May be you can decide better what to parse in there.

**Comment by hunjixin** [comment]
@abhi3700 Please test it on macos

File: Cargo.toml
Comment (nazar-pc) [inline Cargo.toml:39]: This dependency is not needed on Windows and should be in corresponding platform-specific section like Windows dependencies. Also please move newly added dependencies such that everything remains sorted alphabetically.

Also for `image` we only need `png` support, so use `default-features = false, features = ["png"]`, otherwise by default it pulls support for huge number of formats we don't need: https://github.com/image-rs/image/blob/4594c80b3626076af128b7a853c8e5c5187f7a64/Cargo.toml#L70

File: src/main.rs
Comment (nazar-pc) [inline src/main.rs:509]: It is not a PNG icon in Windows resource, it is [ICO](https://en.wikipedia.org/wiki/ICO_(file_format))

File: src/main.rs
Comment (nazar-pc) [inline src/main.rs:87]: Later I'll probably make a copy of this icon in `unix` subdirectory rather than `linux` and make it much smaller. There is no need to have 256x256 image as a tray icon.

**Comment by hunjixin** [comment]
@nazar-pc can this pr ready for merge? or anything to imporve?

**Comment by abhi3700** [comment]
> Waiting for @abhi3700 to confirm macOS is working, I thought that was the remaining issue

It works fine!

**Comment by nazar-pc** [comment]
> There is no build error found on macOS. Also, no clippy issues.

We can see that from CI, does it actually work properly on macOS when you run it?

**Comment by hunjixin** [comment]
@nazar-pc  configure gtk fail,  need i to rebase main branch

**Comment by nazar-pc** [comment]
Indeed, pleae do. Also switch to upstream release of `betrayer` if it is out already and squash everything since changes are not that large here.

**Review by nazar-pc — APPROVED** [review]
Looks good to go, thanks and sorry it took so long to merge it!

**Comment by EmilFattakhov** [comment]
@hunjixin, we would like to thank you once again for your contribution! Please fill out [the compliance form](https://forms.gle/gtjavXSW6nF2wh216) to claim your reward! Looking forward to your future contributions!

**Comment by abhi3700** [comment]
> > There is no build error found on macOS. Also, no clippy issues.
> 
> We can see that from CI, does it actually work properly on macOS when you run it?

Apologise for misunderstanding.

I am not sure how it should look like. Can I have a preview from other OS so as to verify?

**Comment by hunjixin** [comment]
> @hunjixin, we would like to thank you for your contribution! Please fill out [the compliance form](https://forms.gle/gtjavXSW6nF2wh216) to claim your reward! Looking forward to your future contributions!

work is okay,  betray not support on macos, because i and author of betrayer either have no macos, difficulty to complete this work.

**Comment by abhi3700** [comment]
> > @hunjixin, we would like to thank you for your contribution! Please fill out [the compliance form](https://forms.gle/gtjavXSW6nF2wh216) to claim your reward! Looking forward to your future contributions!
> 
> work is okay, betray not support on macos, because i and author of betrayer either have no macos, difficulty to complete this work.

@hunjixin
Are you confirming tray icon is not supported on macOS or just so you don't have access to mac, so it's not checked?

**Comment by hunjixin** [comment]
> > > @hunjixin, we would like to thank you for your contribution! Please fill out [the compliance form](https://forms.gle/gtjavXSW6nF2wh216) to claim your reward! Looking forward to your future contributions!
> > 
> > 
> > work is okay, betray not support on macos, because i and author of betrayer either have no macos, difficulty to complete this work.
> 
> @hunjixin Are you confirming tray icon is not supported on macOS or it's not checked as you don't have mac?

hmm, i see this code for macos https://github.com/sidit77/betrayer/blob/1dbe481a0b9512b3357e3a2ea9cb5ff6edac3b89/src/platform/macos/mod.rs#L98, https://github.com/sidit77/betrayer/blob/1dbe481a0b9512b3357e3a2ea9cb5ff6edac3b89/src/platform/macos/mod.rs#L119-L121. this code do nothing with given icon, so i think macos cannot work with feature.  not sure the menu can work or not.  you can see this https://github.com/subspace/space-acres/issues/142  to know the effection of this pr

**Comment by hunjixin** [comment]
@abhi3700   image

**Comment by nazar-pc** [comment]
Unfortunately this didn't work with my custom X11-based DE (I'm using Qtile WM with tint2 as the panel):
```
2024-04-24T00:36:39.821252Z  WARN space_acres: Unable to create tray icon  err=Os(MethodError(OwnedErrorName("org.freedesktop.DBus.Error.ServiceUnknown"), Some("The name org.kde.StatusNotifierWatcher was not provided by any .service files"), Msg { type: Error, sender: UniqueName("org.freedesktop.DBus"), reply-serial: 7, body: Signature("s"), fds: [] }))
```

We'll likely want to improve or replace `betrayer` to improve compatibility. But at least it still works as expected outside of that.

**Comment by hunjixin** [comment]
@nazar-pc  sorry I'm not familiar with the Qtile system,  but I searched through the Qtile codebase, and it seems that Qtile does support the [org.kde.StatusNotifierWatcher](https://github.com/qtile/qtile/blob/3e7284b0514e47eb700900fcc037812b86b8481e/libqtile/widget/statusnotifier.py#L50) signal,  so I think the support should be fine. but I still need to dig more(on the system and the code).

**Comment by nazar-pc** [comment]
Maybe their bar does, but I don't use it, I use tint2 that has some kind of generic system tray that all other apps work fine with.

**Comment by hunjixin** [comment]
try this tool with ubunut, work fine.  i need to try the tool on qtile.
image

**Comment by nazar-pc** [comment]
I don't think it has anything to do with Qtile specifically, likely Tint2. Run Openbox or similar simple window manager and tint2 and you should be able to reproduce it.

**Comment by abhi3700** [comment]
The tray looks like this on macOS:
image

I don't know where is this "TEST BUTTON" name defined 🤔 . Is it in gtk library or somewhere else?

Secondly,I noticed the tray icon goes away when switched to another application. As you can see here in image below that there is no "TEST BUTTON" when I switched to browser.
image

So, these are 2 issues on macOS related to this PR.

@nazar-pc should we open these issues separately to may be solved later?

**Comment by nazar-pc** [comment]
Just open one issue for macOS tray icon and describe both in there

**Pull request merged**
